sw/device/tests/sim_dv/otp_ctrl_hw_cfg_test.c

```diff
@@ -0,0 +1,97 @@
+#include <stdbool.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_lc_ctrl.h"
+#include "sw/device/lib/dif/dif_otp_ctrl.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "lc_ctrl_regs.h"  // Generated.
+
+static dif_otp_ctrl_t otp;
+static dif_lc_ctrl_t lc;
+
+const test_config_t kTestConfig;
+const uint8_t NUM_DEVICE_ID = 8;
+
+/**
+ * Busy-wait until the DAI is done with whatever operation it is doing.
+ */
+static void wait_for_dai(void) {
+  while (true) {
+    dif_otp_ctrl_status_t status;
+    CHECK(dif_otp_ctrl_get_status(&otp, &status) == kDifOtpCtrlOk);
+    if (bitfield_bit32_read(status.codes, kDifOtpCtrlStatusCodeDaiIdle)) {
+      return;
+    }
+    LOG_INFO("Waiting for DAI...");
+  }
+}
+
+/**
+ * Read and return 32-bit OTP data via the DAI interface.
+ */
+static void otp_ctrl_dai_read_32(const dif_otp_ctrl_t *otp,
+                                 dif_otp_ctrl_partition_t partition,
+                                 uint32_t address, uint32_t *buf) {
+  dif_otp_ctrl_dai_result_t read_start_err =
+      dif_otp_ctrl_dai_read_start(otp, partition, address);
+  CHECK(read_start_err == kDifOtpCtrlDaiOk,
+        "Failed to perform OTP DAI read start.");
+
+  wait_for_dai();
+
+  dif_otp_ctrl_dai_result_t read_end_err =
+      dif_otp_ctrl_dai_read32_end(otp, buf);
+  CHECK(read_end_err == kDifOtpCtrlDaiOk,
+        "Failed to perform OTP DAI read end.");
+}
+
+/**
+ * Tests that the OTP sends correct HW_CFG partition data to the receiving IPs.
+ */
+// TODO: needs to support other recipients besides LC_CTRL.
+bool test_main(void) {
+  mmio_region_t otp_reg =
+      mmio_region_from_addr(TOP_EARLGREY_OTP_CTRL_BASE_ADDR);
+  CHECK(dif_otp_ctrl_init((dif_otp_ctrl_params_t){.base_addr = otp_reg},
+                          &otp) == kDifOtpCtrlOk);
+
+  mmio_region_t lc_reg = mmio_region_from_addr(TOP_EARLGREY_LC_CTRL_BASE_ADDR);
+  CHECK(dif_lc_ctrl_init((dif_lc_ctrl_params_t){.base_addr = lc_reg}, &lc) ==
+        kDifLcCtrlOk);
+
+  dif_otp_ctrl_config_t config = {
+      .check_timeout = 100000,
+      .integrity_period_mask = 0x3ffff,
+      .consistency_period_mask = 0x3ffffff,
+  };
+  CHECK(dif_otp_ctrl_configure(&otp, config) == kDifOtpCtrlLockableOk);
+
+  // Read out Device ID from OTP DAI read, and compare the value with LC_CTRL's
+  // `device_id` registers.
+  // TODO: current HW CFG value is randomly genenrated from the HJSON file,
+  // plan to backdoor inject.
+  uint32_t otp_device_id;
+  for (int i = 0; i < NUM_DEVICE_ID; i++) {
+    otp_ctrl_dai_read_32(&otp, kDifOtpCtrlPartitionHwCfg, i * 4,
+                         &otp_device_id);
+
+    uint32_t lc_device_id = mmio_region_read32(
+        lc.params.base_addr, LC_CTRL_DEVICE_ID_0_REG_OFFSET + i * 4);
```

Comment:
@moidx, @cfrantz 
This DIF is missing. The requirements for LC ctrl testing is documented in the chip testplan here:
https://docs.opentitan.org/hw/top_earlgrey/doc/dv/#dv-plan

If you search for `chip_lc_ctrl`, you can see our testing strategy / requirements for LC ctrl DIFs. If you scroll down to `chip_lc_ctrl_otp_hw_cfg`, it enumerates the steps for that test, which should somewhat clarify DV DIF needs for LC ctrl.

@cindychip 
Do you want to take a stab at writing this DIF? Add a separate commit in this PR itself (make that the first commit) and have that reviewed with SW team.

Reply:
Thanks sure, if SW teammates do not mind, I am happy to give it a try.

Reply:
Hi @cindychip, @cfrantz, SGTM. A separate PR would be ok for this. Thanks!

Reply:
@cindychip sure, I'd be happy to review PRs to create the lifecycle controller DIF.

Reply:
Awesome, thanks I definitely need lots of reviews and suggestions :) Will try it today